0001-typedb/02-dev/02-schema.md

```diff
@@ -0,0 +1,948 @@
+
+<div class="note">
+[Note]
+Roles can be queried like any other types even though roles do not have a direct type definition.
+</div>
+
+### Attribute types
+
+Attribute types represent the properties of `entity` and `relation` types (e.g., the name of a business unit or 
+a user) with a value. In TypeDB, any type can **own** an attribute type. However, different types can own the same 
+attribute type — and different instances can share ownership of the same attribute instance.
+
+For example, multiple users can own the same instance of an attribute type with the value of `Alex`.
+
+<div class="note">
+[Warning]
+The feature of attribute type owning another attribute types will be deprecated in TypeDB version 3.0.
+</div>
+
+### Inheritance
+
+A type can subtype another type. As a result, the subtype inherits all the attributes owned and roles played by its 
+supertype. However, while a type can only have a single supertype, types can be subtypes of other subtypes, 
+resulting in a type hierarchy. For example, `business unit` subtypes `user group`, which subtypes `subject`, 
+which subtypes `entity` base type.
+
+<div class="note">
+[Note]
+While roles are not generally considered to be types, they can be inherited and even overriden as a part of relation 
+inheritance.
+</div>
+
+## Rules
+
+Reasoning engine uses rules as a set of logic to infer new data. A rule consists of a condition and a conclusion. 
+Condition is a pattern to look for in data and conclusion is an action to take for every found pattern-matching piece 
+of data. Rules are used by read queries for 
+[Inferring new data](06-infer.md).
+
+<div class="note">
+[Important]
+Rules can’t change persisted data in a database. Instead, all the reasoning is done within a dataset of a query. 
+Only the result of queries is influenced by the inference. During a single transaction, newly inferred facts will be 
+retained and reused (with corresponding performance gains). New transactions will re-compute inferred facts again.
+</div>
+
+The rules syntax uses `when` and `then` keywords as in `if`/`then` statement that produce virtual relations or 
+attributes when the conditions are met. Read more on how to create rules in a schema in the
+[Define rules](#define-rules) section. See the example of rules syntax below.
+
+<!-- test-ignore -->
+```typeql
+define
+
+rule rule-label:
+when {
+    ## the conditions
+} then {
+    ## the conclusion
+};
+```
+
+<div class="note">
+[Important]
+Inference can only be used in a read transaction. Learn how to use inference on the [Inferring data](06-infer.md) page.
+</div>
+
+<div class="note">
+[Note]
+The approach TypeDB uses is a [backward-chaining](https://en.wikipedia.org/wiki/Backward_chaining) execution on top 
+of [Horn-clause](https://en.wikipedia.org/wiki/Horn_clause) logic. Negation functionality follows the 
+[negation-as-failure](https://en.wikipedia.org/wiki/Negation_as_failure) model. Infinite recursion and non-termination 
```

Comment:
The talk helped me get a better picture.
Negation as failure (NAF) is suited for a more general setting like prolog, where the order in which the predicates are written determines the order of execution and there can be free variables in results.

Our setting (stratified negation & ground results) allows the set-difference semantics. I would say the ideal way to present it is to go with the set-difference semantics, since it's easier to understand. And say it corresponds to NAF under the conditions where :
* we have stratified negation
* the results are grounded
* we ensure all variables occurring both inside and outside the negation are bound by the time the negation is evaluated.

Reply:
I found the source I've been using - https://medium.com/@joshua_73415/e95ba904b7d8 @flyingsilverfin I know it was just a draft. But you might want to also update it accordingly for future publication.

Reply:
Fixed in https://github.com/vaticle/docs/pull/676/commits/9837352d0f221e3dbba830796719cd3cc33f09b6 
But I failed to find any good sources to provide links for the terms set-difference semantics, grounded results. And the stratified negation has some wiki link, but it's kind of rudimental. I'm not sure whether we need to go into such details. 

Can you provide 1-2 or more examples of interesting nuances of how our particular negation works in comparison to what is generally can be anticipated by a spherical developer in vacuum? Or just an interesting example or two with negation.

Reply:
Thank you. Not sure we need a separate page as it was before. Maybe a blog post describing all the details would be a better format. But I'll try to fit some short explanation into our docs.